internal/service/subscription.go

```diff
@@ -3387,3 +3390,226 @@ func (s *subscriptionService) ActivateIncompleteSubscription(ctx context.Context
+		if inv.DueDate == nil {
+			s.Logger.Warnw("invoice has invalid due date, skipping",
+				"subscription_id", sub.ID,
+				"invoice_id", inv.ID)
+			continue
+		}
+
+		// Check amount_remaining (must have outstanding amount)
+		if !inv.AmountRemaining.GreaterThan(decimal.Zero) {
+			s.Logger.Debugw("invoice has no remaining amount, skipping",
+				"subscription_id", sub.ID,
+				"invoice_id", inv.ID,
+				"amount_remaining", inv.AmountRemaining)
+			continue
+		}
+
+		// Calculate grace period end time: due_date + grace_period_days
+		gracePeriodEndTime := inv.DueDate.AddDate(0, 0, config.GracePeriodDays)
+
+		s.Logger.Debugw("evaluating invoice for auto-cancellation",
+			"subscription_id", sub.ID,
+			"invoice_id", inv.ID,
+			"due_date", inv.DueDate,
+			"amount_remaining", inv.AmountRemaining,
+			"grace_period_days", config.GracePeriodDays,
+			"grace_period_end_time", gracePeriodEndTime,
+			"current_time", now,
+			"is_past_grace_period", now.After(gracePeriodEndTime))
+
+		// Check if current time is past grace period end
+		if now.After(gracePeriodEndTime) {
+			s.Logger.Infow("subscription eligible for auto-cancellation",
+				"subscription_id", sub.ID,
+				"invoice_id", inv.ID,
+				"amount_remaining", inv.AmountRemaining,
+				"due_date", inv.DueDate,
+				"grace_period_end_time", gracePeriodEndTime,
+				"current_time", now,
+				"payment_status", inv.PaymentStatus)
+			return true
+		} else {
+			s.Logger.Debugw("invoice not past grace period yet",
+				"subscription_id", sub.ID,
+				"invoice_id", inv.ID,
+				"due_date", inv.DueDate,
+				"grace_period_end_time", gracePeriodEndTime,
+				"days_until_grace_expires", gracePeriodEndTime.Sub(now).Hours()/24)
+		}
+	}
+
+	s.Logger.Debugw("subscription not eligible for auto-cancellation",
+		"subscription_id", sub.ID,
+		"reason", "no invoices past grace period")
+
+	return false
+}
+
+// ProcessAutoCancellationSubscriptions processes subscriptions that are eligible for auto-cancellation
+func (s *subscriptionService) ProcessAutoCancellationSubscriptions(ctx context.Context) error {
+	s.Logger.Infow("starting auto-cancellation processing")
+
+	// Get subscription configuration
+	config, err := s.GetSubscriptionConfig(ctx)
+	if err != nil {
+		s.Logger.Errorw("failed to get subscription config", "error", err)
+		return err
+	}
+
+	// Early return if auto-cancellation is disabled at tenant level
+	if !config.AutoCancellationEnabled {
+		s.Logger.Infow("auto-cancellation is disabled for tenant, skipping",
+			"tenant_id", types.GetTenantID(ctx),
+			"environment_id", types.GetEnvironmentID(ctx))
+		return nil
+	}
+
+	// Get ONLY ACTIVE subscriptions (skip cancelled, paused, trialing, etc.)
+	filter := &types.SubscriptionFilter{
+		SubscriptionStatus: []types.SubscriptionStatus{types.SubscriptionStatusActive},
+	}
+
+	s.Logger.Infow("fetching ONLY ACTIVE subscriptions for auto-cancellation",
+		"subscription_status_filter", "ACTIVE")
+
+	subscriptions, err := s.SubRepo.List(ctx, filter)
```

Comment:
Need to list in a loop because a single tenant can have lots of subscriptions
also need to cut the search space to pick maybe the ones with due date > time now
this will significantly reduce the number of rows being generated.

Reply:
In order to reduce the scope we can again filter and just pull all rows from settings table so that we run this only for tenant x env where the setting has been set